### PR TITLE
After card reset detected, run SM open under new transaction

### PR DESCRIPTION
**Pull request by dengert**
@viktorTarasov
I have questions on the logic used in sc_lock when card->reader->ops->lock(card->reader) returns SC_ERROR_CARD_RESET after a SCardBeginTransaction fails. It appears the calls to sm_ctx.ops.open is in the wrong place and may cause problems if there are many processes trying to use the same card.

sm_ctx.ops.open is called  while the card mutex is held and there is no PCSC transaction place.
The sm_ctx.ops.open can issue APDU, which will do sc_lock and pcsc_lock to do SCardBeginTranasction and when sm_ctx.ops.open retuerns, pcsc_unlock will be called and thus the transaction would have only been done for the sm_ctx.ops.open. Then when sm_ctx.ops.open returns to sc_lock,  sc_lock will will try and obtain a new transaction. But this new transaction could also fail because of SC_ERROR_CARD_RESET but sm_ctx.ops.open.

This PR is an attempt to address the issue. 

What I think should happen, is the original sc_lock should get the new transaction before sm_ctx.ops.open is called  so sm_ctx.ops.open is run on the same transaction as will be used by the original caller of sc_lock.

In other words after a card was reset, the  sm_ctx.ops.open should be run under a newly acquired Transaction and this is the Transaction returned  to the caller.

What appears to happen is sm_ctx.ops.open is run under its own Transaction then a new Transaction is obtained that is returned to the caller. This second Transaction could also return or interfered with by other processes, and it is not handled by the current code.

I do not have any SM cards, so can not test this. But I want to add #836 that would need to be processed just like the SM open.

### CODE REVIEW

**Comment by viktorTarasov** [comment]
>  ... sc_lock when card->reader->ops->lock(card->reader) returns SC_ERROR_CARD_RESET after a SCardBeginTransaction fails.

Afais, _SC_ERROR_CARD_RESET_ is [returned by _pcsc_lock_](https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/reader-pcsc.c#L612) only when card is successfully reconnected and with a valid transaction (after successful SCardBeginTransaction).
So, as for me, _Open SM_ is in its right place, with a valid transaction and within the locked application mutex.

**Comment by viktorTarasov** [comment]
Probably I was a little bit too rapid. I will look it over once more.

**Comment by viktorTarasov** [comment]
@dengert:
I agree with your proposal,
the only thing is that the _Open SM_ has to be done before _sm_mutex_unlock()_ call.

**Comment by dengert** [comment]
You mean the  sc_mutex_unlock?
I will rework the change to keep the sm code before the sc_mutex_unlock.


    On 7/25/2016 7:49 AM, viktorTarasov
      wrote:
    
    
      @dengert:
        I agree with your proposal,
        the only thing is that the Open SM has to be done
        before sm_mutex_unlock() call.
      —
        You are receiving this because you were mentioned.
        Reply to this email directly, view
          it on GitHub, or mute
          the thread.
      
        
          
          
        
        
      
    
    
    -- 

 Douglas E. Engert  <DEEngert@gmail.com>

**Comment by dengert** [comment]
@viktorTarasov 
I pushed a new version of the PR, where sm code os called befire sc_mutex_unlock
